mealie: init at 1.2.0

litchipi opened the pull request:
## Description of changes

Package request: https://github.com/NixOS/nixpkgs/issues/254260

Adds a new package `mealie`, with its NixOS module.

Work mostly inspired from [the work previously made by @fleaz](https://github.com/NixOS/nixpkgs/commit/af76ffabba2ff0f81a5a5cf513f3998bf81af526), and tested on a [Qemu VM setup](https://github.com/litchipi/nixos-service-builder/tree/mealie) before being ported here.

Mealie is a self hosted recipe manager and meal planner with a RestAPI backend and a reactive frontend
application built with NuxtJS for a pleasant user experience for the whole family.
Easily add recipes into your database by providing the URL and Mealie will automatically import the relevant data or add a family recipe with the UI editor.

(See more in [homepage](https://nightly.mealie.io))

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

litchipi commented:
@ambroisie Can't add you as a reviewer, but here's the PR for mealie.
Also @fleaz worked on it and may want to take a look (as 90% of this is his code :stuck_out_tongue: )

Tried to fit Contributing, let me know if you see anything.

Also: **the package wasn't tested outside a systemd service**, I don't really know yet if that would work directly from the `./result/bin`

ambroisie commented on maintainers/maintainer-list.nix:
```diff
@@ -10593,6 +10593,12 @@
     githubId = 591860;
     name = "Lionello Lunesu";
   };
+  litchipi = {
+    email = "litchi.pi@proton.me";
+    github = "litchipi";
+    githubId = 61109829;
+    name = "Litchi Pi";
+  };
```
Please split this in a commit with title `maintainers: add litchipi`.

ambroisie commented on nixos/doc/manual/release-notes/rl-2405.section.md:
```diff
@@ -39,6 +39,8 @@ The pre-existing [services.ankisyncd](#opt-services.ankisyncd.enable) has been m
 
 - [TuxClocker](https://github.com/Lurkki14/tuxclocker), a hardware control and monitoring program. Available as [programs.tuxclocker](#opt-programs.tuxclocker.enable).
 
+- [Mealie](https://nightly.mealie.io/), a self-hosted recipe manager and meal planner with a RestAPI backend and a reactive frontend application built in NuxtJS for a pleasant user experience for the whole family. Available as [services.mealie](#opt-services.mealie.enable)
+
```
Using this change to talk about the commit in general: please just squash it to the `mealie: init at ...` one (for a cleaner history).

If you want  to keep "revisions" separate for reviewing reasons, I'd recommend looking into `git commit --fixup` and `git rebase -i --autosquash` 👍.

litchipi replied:
I cannot squash it on top of the previous work, it was never merged before, so in `master` it's the first `mealie: init...`. Or maybe I didn't understood what you meant ? :sweat_smile:

ambroisie replied:
Sorry, to make my point clearer: IMO this PR should have four commits, in this order:

* `maintainers: add litchipi`: adds yourself to the maintainers list
* `mealie: init at <version>`: adds the package
* `nixos/mealie: init module`: add the module
* `nixos/mealie: init test`: test the moduel, and add `passthru.tests = { inherit (nixosTests) mealie; }` to the package, for CI purposes.

Usually, the documentation for a new module is added in the same commit as the module's creation, I usually add it in a separate commit at the end: do as you prefer.

ambroisie commented on nixos/modules/services/web-apps/mealie.nix:
```diff
@@ -0,0 +1,108 @@
+    host = lib.mkOption {
+      type = lib.types.str;
+      default = "localhost";
+      description = "Host on which creating the URL on the service";
```
Is it a listen address? Or for links created on the UI?

litchipi replied:
Only links for the UI

ambroisie commented on nixos/modules/services/web-apps/mealie.nix:
```diff
@@ -0,0 +1,108 @@
+    protocol = lib.mkOption {
+      type = lib.types.str;
+      default = "http";
+      description = "Protocol to use to serve the service";
```
Same question, I assume this is for the links on the UI?

ambroisie commented on nixos/modules/services/web-apps/mealie.nix:
```diff
@@ -0,0 +1,108 @@
+    };
+
+    logLevel = lib.mkOption {
+      type = lib.types.str;
```
`lib.types.enum [ "INFO" ... ]`?

Not a strong review, I could go either way.

ambroisie commented on nixos/modules/services/web-apps/mealie.nix:
```diff
@@ -0,0 +1,108 @@
+        User = "mealie";
+        ExecStartPre = let
+          exec = pkgs.writeShellScript "startup-mealie.sh" ''
+            ${pkgs.toybox}/bin/sed 's+script_location = alembic+script_location = ${pkg.src}/alembic+g' ${pkg.src}/alembic.ini > $ALEMBIC_CONFIG_FPATH
```
I don't see a strong reason to use `toybox` , I'd expect `gnused` to already be in the system closure anyway.


 ```suggestion
            ${lib.getExe pkgs.gnused} 's+script_location = alembic+script_location = ${pkg.src}/alembic+g' ${pkg.src}/alembic.ini > $ALEMBIC_CONFIG_FPATH
```

litchipi replied:
Indeed, that was a dirty fix for testing purpose

ambroisie commented on nixos/modules/services/web-apps/mealie.nix:
```diff
@@ -0,0 +1,108 @@
+            ${pkgs.toybox}/bin/sed 's+script_location = alembic+script_location = ${pkg.src}/alembic+g' ${pkg.src}/alembic.ini > $ALEMBIC_CONFIG_FPATH
+            ${pkg.interpreter} ${pkg}/lib/${pkg.python.libPrefix}/site-packages/mealie/db/init_db.py
+          '';
+        in "${exec}";
```
I'd remove the redundant `let ... in` block here.

ambroisie commented on nixos/modules/services/web-apps/mealie.nix:
```diff
@@ -0,0 +1,108 @@
+      ];
+
+      environment = {
+        PYTHONPATH = "${pkg.pythonPath}:${pkg}/lib/${pkg.python.libPrefix}/site-packages";
```
Rather than set `PYTHONPATH`, I'd recommend you take a look at `tandoor-recipes` and `paperperless`'s way of invoking `gunicorn`.

This feels more brittle in comparison.

Let me know if you need more direction.

ambroisie commented on nixos/modules/services/web-apps/mealie.nix:
```diff
@@ -0,0 +1,108 @@
+        ALEMBIC_CONFIG_FPATH="/var/lib/mealie/alembic.ini";
+
+        # Default configs
+        DEFAULT_GROUP="Home";
+        DEFAULT_EMAIL="changeme@example.com";
+        DEFAULT_PASSWORD="MyPassword";
```
Is that the default user's configuration?

If so, I would add an option e.g: `defaultUserConfigFile` to override this, rather than just use bad defaults.

litchipi replied:
Yeah I just copied the launching settings from the sources of `mealie`.
Should I create an option taking an `attrsets` to set these ?
I saw in the documentation that it's possible to use a `.env` file, but it has to be located somewhere specific, and because of all the "paths magic" (links to nix store, frontend not at the same place, etc...) I didn't really bother to use this.
Is that what should be done here ?

ambroisie commented on nixos/modules/services/web-apps/mealie.nix:
```diff
@@ -0,0 +1,108 @@
+        DEFAULT_EMAIL="changeme@example.com";
+        DEFAULT_PASSWORD="MyPassword";
+        PRODUCTION = "true";
+        API_DOCS = "False";
+        DB_ENGINE = "sqlite";
+        TOKEN_TIME="24";
+
+        # Configuration from module
+        LOG_LEVEL = cfg.logLevel;
+        API_PORT = builtins.toString cfg.port;
+        BASE_URL = "${cfg.protocol}://${cfg.host}:${builtins.toString cfg.port}";
+
+        # TODO  Additionnal config
+        # ALLOW_SIGNUP = cfg.allow_signup;
+        # POSTGRES_USER=mealie
+        # POSTGRES_PASSWORD=mealie
+        # POSTGRES_SERVER=postgres
+        # POSTGRES_PORT=5432
+        # POSTGRES_DB=mealie
+        # LDAP_AUTH_ENABLED=False
+        # LDAP_SERVER_URL=""
+        # LDAP_TLS_INSECURE=False
+        # LDAP_TLS_CACERTFILE=
+        # LDAP_ENABLE_STARTTLS=False
+        # LDAP_BASE_DN=""
+        # LDAP_QUERY_BIND=""
+        # LDAP_QUERY_PASSWORD=""
+        # LDAP_USER_FILTER="(&(|({id_attribute}={input})({mail_attribute}={input}))(objectClass=person))"
+
+        # LDAP_ADMIN_FILTER=""
+        # LDAP_ID_ATTRIBUTE=uid
+        # LDAP_NAME_ATTRIBUTE=name
+        # LDAP_MAIL_ATTRIBUTE=mail
```
Rather than try to introduce 1001 options, I would add an `extraConfig` option similar to `paperless`.

You can them do something like:

```
services.mealie.extraConfig = {
    DB_ENGINE = lib.mkDefault "sqlite";
    API_DOCS = lib.mkDefault "False";
    TOKEN_TIME = lib.mkDefault "24";
    PRODUCTION = lib.mkDefault "true";
    # Arguably, `logLevel` isn't very useful as an option, it could just 
    LOG_LEVEL = cfg.logLevel;
};
```

BTW: different casing between `False` and `true` looks weird.

litchipi replied:
Ok I tried to do it like this, it should be much better now !
Also, I noticed the case difference between booleans, but frankly I'm too scared to attempt to put everything in a case, as it may screw up env var "parsing" inside the Python code

ambroisie commented on nixos/modules/services/web-apps/mealie.nix:
```diff
@@ -0,0 +1,108 @@
+            ${pkg.interpreter} ${pkg}/lib/${pkg.python.libPrefix}/site-packages/mealie/db/init_db.py
+          '';
+        in "${exec}";
+        ExecStart = "${pkg.pythonPkg.gunicorn}/bin/gunicorn -b 0.0.0.0:${builtins.toString cfg.port} -k uvicorn.workers.UvicornWorker mealie.app:app";
```
The listen address should be configurable.

ambroisie commented on pkgs/servers/web-apps/mealie/default.nix:
```diff
@@ -0,0 +1,123 @@
+{ python310, python310Packages, fetchFromGitHub, mkYarnPackage, lib, ...}: let
+  version = "v1.0.0-RC2";
+  src = fetchFromGitHub {
+    owner = "mealie-recipes";
+    repo = "mealie";
+    rev = version;
+    sha256 = "sha256-/sht8s0Nap6TdYxAqamKj/HGGV21/8eYCuYzpWXRJCE=";
+  };
```
```suggestion
  version = "1.0.0-RC2";
  
  src = fetchFromGitHub {
    owner = "mealie-recipes";
    repo = "mealie";
    rev = "v{version}";
    sha256 = "sha256-/sht8s0Nap6TdYxAqamKj/HGGV21/8eYCuYzpWXRJCE=";
  };
```

ambroisie commented on pkgs/servers/web-apps/mealie/default.nix:
```diff
@@ -0,0 +1,123 @@
+    sha256 = "sha256-/sht8s0Nap6TdYxAqamKj/HGGV21/8eYCuYzpWXRJCE=";
+  };
+
+  frontend = (mkYarnPackage {
+    name = "mealie-frontend";
+    src = "${src}/frontend";
+    packageJSON = "${src}/frontend/package.json";
+    yarnLock = "${src}/frontend/yarn.lock";
+  }).overrideAttrs (old:
+    {
+      doDist = false;
+
+      buildPhase = ''
+        runHook preBuild
+        shopt -s dotglob
+
+        rm deps/${old.pname}/node_modules
+        mkdir deps/${old.pname}/node_modules
+        pushd deps/${old.pname}/node_modules
+        ln -s ../../../node_modules/* .
+        popd
+        export NUXT_TELEMETRY_DISABLED=1
+        yarn --offline --non-interactive --production=false build
+        yarn --offline --non-interactive --produciton=false generate
+        runHook postBuild
+      '';
+
+      installPhase = ''
+        runHook preInstall
+        mv deps/${old.pname}/dist $out
+        runHook postInstall
+      '';
+    });
```
At the very least, you shouldn't need `overrideAttrs` (my indentation might be wrong)

```suggestion
  frontend = mkYarnPackage {
    name = "mealie-frontend";
    src = "${src}/frontend";

    packageJSON = "${src}/frontend/package.json";
    yarnLock = "${src}/frontend/yarn.lock";

    doDist = false;

    buildPhase = ''
      runHook preBuild
      shopt -s dotglob

      rm deps/${old.pname}/node_modules
      mkdir deps/${old.pname}/node_modules
      pushd deps/${old.pname}/node_modules
      ln -s ../../../node_modules/* .
      popd
      export NUXT_TELEMETRY_DISABLED=1
      yarn --offline --non-interactive --production=false build
      yarn --offline --non-interactive --produciton=false generate

      runHook postBuild
    '';

    installPhase = ''
      runHook preInstall

      mv deps/${old.pname}/dist $out
    
       runHook postInstall
      '';
    };
```

Secondly: you need to give `yarnOfflineCache = fetchYarnDeps { ... }` (refer to e.g: `tandoor-recipes` for example). I think `mkYarnPackage` will result in IFD otherwise, which is forbidden in nixpkgs.

Thirdly: what's happening with `node_module` in `buildPhase`?

litchipi replied:
I honestly cannot explain you what happens here, I just figured out a way to get my ReactJS / VueJS compilation working in Nix (as I didn't have any `main.js` output using the base setup), and this abomination was the result.
Maybe because I don't know JS ecosystem at all, just trying to build the apps :sweat_smile: 

I looked at the implementation done in `tandoor-recipes` and that seams to work well !

ambroisie commented on pkgs/servers/web-apps/mealie/default.nix:
```diff
@@ -0,0 +1,123 @@
+  format = "pyproject";
+
+  makeWrapperArgs = [
+    "--set STATIC_FILES ${frontend}"
```
If it's already here, no need for it in the systemd service.

ambroisie commented on pkgs/servers/web-apps/mealie/default.nix:
```diff
@@ -0,0 +1,123 @@
+    pyyaml
+    sqlalchemy
+    aiofiles
+    # Outdated one in NixOS package repository
+    (alembic.overrideAttrs (old: {
+      src = fetchFromGitHub {
+        owner = "sqlalchemy";
+        repo = "alembic";
+        rev = "rel_1_11_3";
+        sha256 = "sha256-AmkLfrmTu5KLILDBVqLo6on7xtT8hSSnxnNym6dzEq4=";
+      };
+    }))
```
Unless it's a breaking change for other packages, you should update the `alembic` instead (in a different commit, prior to this one).

If it would otherwise work with the version of `alembic` currently packaged (i.e: it's just failing because the lock file is too restrictive in its version requirements), just use [`pythonRelaxDepsHook`](https://nixos.org/manual/nixpkgs/unstable/#using-pythonrelaxdepshook).

ambroisie commented on pkgs/servers/web-apps/mealie/default.nix:
```diff
@@ -0,0 +1,123 @@
+  ];
+
+  doCheck = true;
+  checkInputs = with python310Packages; [
+    pytestCheckHook
+  ];
```
```suggestion

  checkInputs = with python310Packages; [
    pytestCheckHook
  ];
```

ambroisie commented on pkgs/servers/web-apps/mealie/default.nix:
```diff
@@ -0,0 +1,123 @@
+    python-jose
+    ldap
+    pytesseract
+  ];
```
Please sort the list ASCII-betically.

ambroisie commented on pkgs/servers/web-apps/mealie/default.nix:
```diff
@@ -0,0 +1,123 @@
+  ];
+
+  passthru = {
+    python = python310;
+    pythonPkg = python310Packages;
+    interpreter = "${python310}/bin/python3";
+    pythonPath = python310.pkgs.makePythonPath propagatedBuildInputs;
+    inherit frontend src;
```
```suggestion
    inherit python frontend src;
```

The other attributes _shouldn't_ be necessary, no need to put them in `passthru`.

ambroisie commented on pkgs/servers/web-apps/mealie/default.nix:
```diff
@@ -0,0 +1,123 @@
+{ python310, python310Packages, fetchFromGitHub, mkYarnPackage, lib, ...}: let
```
Use the non-versioned `python3` please:

Also it is customary to put `lib` as the first argument, usually ASCII-betical attributes after that, and no `...` in packages:

```suggestion
{ lib, fetchFromGitHub, mkYarnPackage, python3, python3Packages }: let
```

ambroisie commented on pkgs/top-level/all-packages.nix:
```diff
@@ -5973,6 +5973,8 @@ with pkgs;
 
   maxcso = callPackage ../tools/archivers/maxcso { };
 
+  mealie = callPackage ../servers/web-apps/mealie { };
+
```
Please package it in `pkgs/by-name`, which is preferred for new packages.

ambroisie reviewed (changes requested):
Took a first look at the module and package.

litchipi commented:
@ambroisie Thank you for the detailed review, tried to address it the best I could.
I'll rebase / squash everything in clean commits once nobody has any more comment, and the whole thing is ready to merge

litchipi commented:
@ambroisie Made all the changes, CI is green, ready for a new round of review ! :slightly_smiling_face:

ambroisie commented on nixos/modules/services/web-apps/mealie.nix:
```diff
@@ -0,0 +1,145 @@
+      API_DOCS = lib.mkDefault "False";
+
+      TOKEN_TIME = lib.mkDefault 24;
+      TZ = lib.mkDefault "UTC";
```
This should be:

```suggestion
      TZ = lib.mkDefault config.time.timeZone;
```

But only if `config.time.timeZone` is non-null.

I would suggest taking inspiration from [the paperless module](https://github.com/NixOS/nixpkgs/blob/bd645e8668ec6612439a9ee7e71f7eac4099d4f6/nixos/modules/services/misc/paperless.nix#L16-L29) for this. So do something like:

```nix
systemd.services.mealie.environment = {
  ...
} // (optionalAttrs config.time.timeZone != null) {
  TZ = config.time.timeZone;
} // (builtins.mapAttrs (_: val: builtins.toString val) cfg.extraConfig);

ambroisie commented on nixos/modules/services/web-apps/mealie.nix:
```diff
@@ -0,0 +1,145 @@
+      DB_ENGINE = lib.mkDefault "sqlite";
+      SECURITY_MAX_LOGIN_ATTEMPTS = lib.mkDefault "5";
+      SECURITY_USER_LOCKOUT_TIME = lib.mkDefault "24";
+
+      POSTGRES_USER = lib.mkDefault "mealie";
+      POSTGRES_PASSWORD = lib.mkDefault "mealie";
+      POSTGRES_SERVER = lib.mkDefault "postgres";
+      POSTGRES_PORT = lib.mkDefault 5432;
+      POSTGRES_DB = lib.mkDefault "mealie";
+
+      SMTP_PORT = lib.mkDefault 587;
+      SMTP_FROM_NAME = lib.mkDefault "mealie";
+      SMTP_AUTH_STRATEGY = lib.mkDefault "TLS";
+
+      WEB_GUNICORN = lib.mkDefault "false";
+      WORKERS_PER_CORE = lib.mkDefault 1;
+      MAX_WORKERS = lib.mkDefault 1;
+      WEB_CONCURRENCY = lib.mkDefault 1;
+
+      LDAP_AUTH_ENABLED = lib.mkDefault "False";
+      LDAP_TLS_INSECURE = lib.mkDefault "False";
+      LDAP_ENABLE_STARTTLS = lib.mkDefault "False";
+      LDAP_ID_ATTRIBUTE = lib.mkDefault "uid";
+      LDAP_NAME_ATTRIBUTE = lib.mkDefault "name";
+      LDAP_MAIL_ATTRIBUTE = lib.mkDefault "mail";
+
+      THEME_LIGHT_PRIMARY = lib.mkDefault "#E58325";
+      THEME_LIGHT_ACCENT = lib.mkDefault "#007A99";
+      THEME_LIGHT_SECONDARY = lib.mkDefault "#973542";
+      THEME_LIGHT_SUCCESS = lib.mkDefault "#43A047";
+      THEME_LIGHT_INFO = lib.mkDefault "#1976D2";
+      THEME_LIGHT_WARNING = lib.mkDefault "#FF6D00";
+      THEME_LIGHT_ERROR = lib.mkDefault "#EF5350";
+      THEME_DARK_PRIMARY = lib.mkDefault "#E58325";
+      THEME_DARK_ACCENT = lib.mkDefault "#007A99";
+      THEME_DARK_SECONDARY = lib.mkDefault "#973542";
+      THEME_DARK_SUCCESS = lib.mkDefault "#43A047";
+      THEME_DARK_INFO = lib.mkDefault "#1976D2";
+      THEME_DARK_WARNING = lib.mkDefault "#FF6D00";
+      THEME_DARK_ERROR = lib.mkDefault "#EF5350";
```
That _way_ too many things, especially since they have default values in the app.

```suggestion
```

ambroisie commented on nixos/modules/services/web-apps/mealie.nix:
```diff
@@ -0,0 +1,145 @@
+      ];
+
+      environment = {
+        PYTHONPATH = "${pkg.pythonPkg.pkgs.makePythonPath pkg.propagatedBuildInputs}:${pkg}/lib/mealie"; #${pkg.python.libPrefix}/site-packages";
```
Please remove the dead comment.

ambroisie commented on nixos/modules/services/web-apps/mealie.nix:
```diff
@@ -0,0 +1,145 @@
+      description = "Host on which the server must listen";
+    };
+
+    url_host = lib.mkOption {
+      type = lib.types.str;
+      default = "localhost";
+      description = "Host on which creating the URL on the service";
+    };
+
```
Looks to be unused.
```suggestion
```

ambroisie commented on nixos/modules/services/web-apps/mealie.nix:
```diff
@@ -0,0 +1,145 @@
+        # Configuration from module
+        LOG_LEVEL = cfg.logLevel;
+        API_PORT = builtins.toString cfg.port;
+        BASE_URL = "${cfg.protocol}://${cfg.host}:${builtins.toString cfg.port}";
```
What if I wanted to have a `BASE_URL` without a port (for example if I have a reverse proxy in front of my service).

Instead of having `protocol`, `host`, `port`, `url_host`, etc... I would have those options:

* `port` and `listenAddress`, which configure `gunicorn` (`-b ${cfg.listenAddress}:${builtins.toString cfg.port}`).
* `BASE_URL` can be configured through `extraConfig` (the same argument holds for `logLevel` as well...).

So actually, I would remove the `BASE_URL` and `LOG_LEVEL` options (`mealie` already has defaults for them).

ambroisie commented on nixos/modules/services/web-apps/mealie.nix:
```diff
@@ -0,0 +1,145 @@
+
+  config = lib.mkIf cfg.enable {
+    services.mealie.extraConfig = {
+      DEFAULT_GROUP = lib.mkDefault "Home";
+      DEFAULT_EMAIL = lib.mkDefault "changeme@example.com";
+      DEFAULT_PASSWORD = lib.mkDefault "MyPassword";
+      ALLOW_SIGNUP = lib.mkDefault "true";
+
+      API_DOCS = lib.mkDefault "False";
+
+      TOKEN_TIME = lib.mkDefault 24;
```
I also think those aren't really necessary, as they have default values in the app. We should really only define the ones that _need_ to be defined by the module, the rest should be taken care of by `extraConfig` if the user wants to override them.

ambroisie commented on nixos/modules/services/web-apps/mealie.nix:
```diff
@@ -0,0 +1,145 @@
+        DynamicUser = true;
+        User = "mealie";
+        ExecStartPre = pkgs.writeShellScript "startup-mealie.sh" ''
+          ${lib.getExe pkgs.gnused} 's+script_location = alembic+script_location = ${pkg.src}/alembic+g' ${pkg.src}/alembic.ini > $ALEMBIC_CONFIG_FPATH
```
Rather than be a `sed` invocation in the service, this patch should be done in the package itself (e.g: in the `postPatch` phase) and the ini packaged up somewhere in `$out` (I don't have a good idea for the exact location yet, I'll think about it).

ambroisie commented on nixos/modules/services/web-apps/mealie.nix:
```diff
@@ -0,0 +1,145 @@
+        User = "mealie";
+        ExecStartPre = pkgs.writeShellScript "startup-mealie.sh" ''
+          ${lib.getExe pkgs.gnused} 's+script_location = alembic+script_location = ${pkg.src}/alembic+g' ${pkg.src}/alembic.ini > $ALEMBIC_CONFIG_FPATH
+          ${pkg.interpreter} ${pkg}/lib/${pkg.python.libPrefix}/site-packages/mealie/db/init_db.py
```
Similarly, can we package this in `$out/bin/init_db` in the package, rather than reach into `site-packages` in the service?

ambroisie commented on pkgs/by-name/me/mealie/mealie-frontend.nix:
```diff
@@ -0,0 +1,73 @@
+    runHook postInstall
+  '';
+}
+
+#   packageJSON = "${src}/frontend/package.json";
+#   yarnLock = "${src}/frontend/yarn.lock";
+
+#   doDist = false;
+
+#   buildPhase = ''
+#     runHook preBuild
+#     shopt -s dotglob
+
+#     rm deps/mealie-frontend/node_modules
+#     mkdir deps/mealie-frontend/node_modules
+#     pushd deps/mealie-frontend/node_modules
+#     ln -s ../../../node_modules/* .
+#     popd
+#     export NUXT_TELEMETRY_DISABLED=1
+#     yarn --offline --non-interactive --production=false build
+#     yarn --offline --non-interactive --produciton=false generate
+#     runHook postBuild
+#   '';
+
+#   installPhase = ''
+#     runHook preInstall
+#     mv deps/mealie-frontend/dist $out
+#     runHook postInstall
+#   '';
+# }
```
Dead code.

ambroisie commented on nixos/modules/services/web-apps/mealie.nix:
```diff
@@ -0,0 +1,145 @@
+        ExecStart = "${pkg.pythonPkg.gunicorn}/bin/gunicorn -b ${cfg.host}:${builtins.toString cfg.port} -k uvicorn.workers.UvicornWorker mealie.app:app";
+        StateDirectory = "mealie";
+        LogsDirectory = "mealie";
+      };
```
One thing which is missing is a `credentialsFile` option, I will refer you to [my pyload PR](https://github.com/NixOS/nixpkgs/pull/276547) as an example, but there should be plenty more in various modules.

mweinelt replied:
This seems unresolved. Is this needed?

anoadragon453 replied:
A `credentialsFile` option was added in [`4ebf2b5` (#278454)](https://github.com/NixOS/nixpkgs/pull/278454/commits/4ebf2b54b09589e35eccb1a565bfb124cb7d09ba#diff-2e9f7142b77cb4eedc4f5be6abaeb7b40f92fdfe8f23ea9b262338f3e402f86aR39).

ambroisie reviewed (changes requested):
It's starting to look very good, please bear with me for some additional changes :-)

fleaz commented:
Thank you @litchipi for putting so much more work into this than I had! 🚀 

And also a big thanks to @ambroisie for the detailed review! 🙏🏻

I will take the code and deploy it on one of my servers on the weekend, and try to spot any problems with it.

Janik-Haag commented:
maintainers entry lgtm; @ambroisie feel free to ping me for a merger once this is ready.

litchipi commented:
@ambroisie Thanks for the review, the coded looks so much cleaner now !
I managed to get it worked with most of the work done in the package building as you asked, and it's really cleaner on the nixos module side.
The whole thing looks neat now, and tests are passing on my side, so I squashed the whole thing into 2 clean commits, ready to merge if you're OK with it :smiley:

fleaz commented:
Just deployed this on my server and can confirm it's working as expected 👍🏻 

* I imported a dump from my previous Mealie installation into it
* Tested the webscraper with some new recipes
* Created/Edited some Users

ambroisie reviewed (changes requested):
@litchipi last nitpick, I promise :-)

Could you split the package, module, and NixOS test into different commits please?

See [this comment](https://github.com/NixOS/nixpkgs/pull/278454/files#r1441670617) for a bit more explanation.

ambroisie commented on nixos/modules/services/web-apps/mealie.nix:
```diff
@@ -0,0 +1,77 @@
+  options.services.mealie = {
+    enable = lib.mkEnableOption "Mealie, a recipe manager and meal planner";
+
+    listen_address = lib.mkOption {
```
Usually, NixOS options are in camel case.

```suggestion
    listenAddress = lib.mkOption {
```

litchipi committed:
maintainers: add litchipi

Signed-off-by: Litchi Pi <litchi.pi@proton.me>

litchipi commented:
@ambroisie No worries, as it's my first contribution, I prefer that you really look closely at what I attempt to merge :sweat_smile:
@fleaz I am having some issues regarding ingredient parsing, I'm gonna look at this, it seams like a missing dependency

litchipi commented:
I can't really make the NLP processing work on ingredients. Idk if that's because it's too "early" in its development.
In case it's really problematic, should we forget about it and remove `crfpp` dependency ? Or should we leave the thing as is ?

ambroisie commented:
@litchipi does it fail hard? Or is it a soft fail surfaced in the UI but otherwise not bringing down the service? If the latter, I'd say it's fine to leave it as a TODO for a future PR.

It might just be that this is a release candidate and not the proper 1.0

fleaz commented:
> @litchipi does it fail hard? Or is it a soft fail surfaced in the UI but otherwise not bringing down the service? If the latter, I'd say it's fine to leave it as a TODO for a future PR.

In the normal usage of Mealie it's not really noticeable (see my previous comment where I tested the app without noticing it :D )
You "just" get a stacktrace in the journal when hitting a place in the app where Mealie wants to use it

litchipi commented:
@ambroisie Yeah what happens is that in the web page, you don't have any result, and in the journal you get a Python stacktrace.
And it's only the NLP detection of ingredients quantities, so pretty niche.
I'll withdraw my attempts to fix it for now, leave a TODO, and I may work on it later on !

ambroisie commented on pkgs/by-name/me/mealie/package.nix:
```diff
@@ -0,0 +1,100 @@
+    '';
+  in ''
+    mkdir -p $out/config $out/bin
+    ${lib.getExe gnused} 's+script_location = alembic+script_location = ${src}/alembic+g' ${src}/alembic.ini > $out/config/alembic.ini
```
Use [`substitute`](https://nixos.org/manual/nixpkgs/unstable/#fun-substitute)/[`substituteInPlace`](https://nixos.org/manual/nixpkgs/unstable/#fun-substituteInPlace) instead of `sed`:

```suggestion
    substitute ${src}/alembic.ini $out/config/alembic.ini \
      --replace 'script_location = alembic' 'script_location = ${src}/alembic'
```

ambroisie commented on pkgs/by-name/me/mealie/package.nix:
```diff
@@ -0,0 +1,100 @@
+
+  passthru = {
+    inherit python3;
+    inherit (nixosTests) mealie;
```
```suggestion
    tests = {
      inherit (nixosTests) mealie;
    };
```

litchipi commented:
(Applied changes and squashed into clean commits)

ambroisie reviewed (approved):
Looks good enough to merge IMO, well done @litchipi :-)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/1390

SuperSandro2000 commented on pkgs/by-name/me/mealie/package.nix:
```diff
@@ -0,0 +1,103 @@
+  ];
+
+  nativeBuildInputs = [
+    python3Packages.poetry-core
+    python3.pkgs.pythonRelaxDepsHook
```
You do not want to mix python3.pkgs and python3Packages as they could point to different pythons if overwrites are done incorrect, resulting in strange bugs and mismatches. Please choose either.

SuperSandro2000 commented on pkgs/by-name/me/mealie/package.nix:
```diff
@@ -0,0 +1,103 @@
+  postInstall = let
+    start_script = writeShellScript "start-mealie" ''
+      export STATIC_FILES="${frontend}"
+      ${python3Packages.gunicorn}/bin/gunicorn "$@" -k uvicorn.workers.UvicornWorker mealie.app:app;
```
```suggestion
      ${lib.getExe python3Packages.gunicorn} "$@" -k uvicorn.workers.UvicornWorker mealie.app:app;
```

SuperSandro2000 commented on pkgs/by-name/me/mealie/package.nix:
```diff
@@ -0,0 +1,103 @@
+      --replace 'script_location = alembic' 'script_location = ${src}/alembic'
+
+    rm -f $out/bin/*
+    ln -s $out/lib/${python3.libPrefix}/site-packages/mealie/db/init_db.py $out/bin/init_db.py
```
```suggestion
    ln -s $out/${python3.sitePackages}/mealie/db/init_db.py $out/bin/init_db.py
```

litchipi commented:
Applied changes and squashed :-)

ambroisie commented on nixos/modules/services/web-apps/mealie.nix:
```diff
@@ -0,0 +1,79 @@
+      default = null;
+      example = "/run/secrets/mealie-credentials.env";
+      description = ''
+        File containing credentials used in mealie such as POSTGRES_PASSWORD
+        or sensitive LDAP options.
+
+        Expects the format of an EnvironmentFile=, as described by systemd.exec(5).
```
```suggestion
        File containing credentials used in mealie such as {env}`POSTGRES_PASSWORD`
        or sensitive LDAP options.

        Expects the format of an `EnvironmentFile=`, as described by {manpage}`systemd.exec(5)`.
```

ambroisie commented on nixos/modules/services/web-apps/mealie.nix:
```diff
@@ -0,0 +1,79 @@
+    listenAddress = lib.mkOption {
+      type = lib.types.str;
+      default = "0.0.0.0";
+      description = "Address on which the service should listen";
```
Nitpick:

```suggestion
      description = "Address on which the service should listen.";
```

ambroisie commented on nixos/modules/services/web-apps/mealie.nix:
```diff
@@ -0,0 +1,79 @@
+    port = lib.mkOption {
+      type = lib.types.port;
+      default = 9000;
+      description = "Port on which to serve the Mealie service";
```
Nitpick:

```suggestion
      description = "Port on which to serve the Mealie service.";
```

ambroisie commented on nixos/modules/services/web-apps/mealie.nix:
```diff
@@ -0,0 +1,79 @@
+      ];
+
+      environment = {
+        PYTHONPATH = "${pkg.python3.pkgs.makePythonPath pkg.propagatedBuildInputs}:${pkg}/lib/${pkg.python3.libPrefix}/site-packages";
```
I think you missed that one:

```suggestion
        PYTHONPATH = "${pkg.python3.pkgs.makePythonPath pkg.propagatedBuildInputs}:${pkg}/${pkg.python3.sitePackages}";
```

And I wonder if perhaps this would also work, as I think it might be slightly more readable:

```suggestion
        PYTHONPATH = "${pkg.python3.pkgs.makePythonPath ([ pkg ] ++ pkg.propagatedBuildInputs)}";
```

ambroisie reviewed (changes requested):
Just a couple more comments, sorry about that

litchipi commented:
Squashed and rebased over (latest) `master` branch !
And no worries @ambroisie , your high quality review makes a high quality PR :smiley:

ambroisie reviewed (approved):
LGTM.

ambroisie commented:
@SuperSandro2000 your review feedback has been addressed, I think we can go for a merge

@Janik-Haag 🏓.

Janik-Haag commented:
for me the build fails locally with:
> adding 'mealie-1.0.0b0.dist-info/WHEEL'
adding 'mealie-1.0.0b0.dist-info/entry_points.txt'
adding 'mealie-1.0.0b0.dist-info/RECORD'
OK
/build/source
Finished executing pypaBuildPhase
Running phase: pythonRuntimeDepsCheckHook
Executing pythonRuntimeDepsCheck
Checking runtime dependencies for mealie-1.0.0b0-py3-none-any.whl
>  - tzdata not installed

and ofborg says the vm tets are failing :/

ambroisie commented:
Weird, the CI is green. I'll try it locally.

ambroisie commented:
Adding `tzdata` to `propagatedBuildInputs` locally, I get another error when trying to build the package:

<details>

```
Running phase: pytestCheckPhase
Executing pytestCheckPhase
ImportError while loading conftest '/build/source/tests/conftest.py'.
tests/conftest.py:13: in <module>
    from mealie.app import app
mealie/app.py:6: in <module>
    from mealie.core.config import get_app_settings
mealie/core/config.py:7: in <module>
    from mealie.core.settings import app_settings_constructor
mealie/core/settings/__init__.py:2: in <module>
    from .settings import *
mealie/core/settings/settings.py:4: in <module>
    from pydantic import BaseSettings, NoneStr, validator
/nix/store/7x4przi4fv2v8z6ywkp798jknr5q18gz-python3.11-pydantic-2.5.2/lib/python3.11/site-packages/pydantic/__init__.py:363: in __getattr__
    return _getattr_migration(attr_name)
/nix/store/7x4przi4fv2v8z6ywkp798jknr5q18gz-python3.11-pydantic-2.5.2/lib/python3.11/site-packages/pydantic/_migration.py:296: in wrapper
    raise PydanticImportError(
E   pydantic.errors.PydanticImportError: `BaseSettings` has been moved to the `pydantic-settings` package. See https://docs.pydantic.dev/2.5/migration/#basesettings-has-moved-to-pydantic-settings for more details.
E
E   For further information visit https://errors.pydantic.dev/2.5/u/import-error
/nix/store/bknngadwym46j65qs14ic2w79rpav888-stdenv-linux/setup: line 1582: pop_var_context: head of shell_variables not a function context
error: builder for '/nix/store/qva7g1gcnyicd09aad6x0liklsc1680r-python3.11-mealie-1.0.0-RC2.drv' failed with exit code 4;
       last 10 log lines:
       > mealie/core/settings/settings.py:4: in <module>
       >     from pydantic import BaseSettings, NoneStr, validator
       > /nix/store/7x4przi4fv2v8z6ywkp798jknr5q18gz-python3.11-pydantic-2.5.2/lib/python3.11/site-packages/pydantic/__init__.py:363: in __getattr__
       >     return _getattr_migration(attr_name)
       > /nix/store/7x4przi4fv2v8z6ywkp798jknr5q18gz-python3.11-pydantic-2.5.2/lib/python3.11/site-packages/pydantic/_migration.py:296: in wrapper
       >     raise PydanticImportError(
       > E   pydantic.errors.PydanticImportError: `BaseSettings` has been moved to the `pydantic-settings` package. See https://docs.pydantic.dev/2.5/migration/#basesettings-has-moved-to-pydantic-settings for more details.
       > E
       > E   For further information visit https://errors.pydantic.dev/2.5/u/import-error
       > /nix/store/bknngadwym46j65qs14ic2w79rpav888-stdenv-linux/setup: line 1582: pop_var_context: head of shell_variables not a function context
       For full logs, run 'nix-store -l /nix/store/qva7g1gcnyicd09aad6x0liklsc1680r-python3.11-mealie-1.0.0-RC2.drv'.
error: 1 dependencies of derivation '/nix/store/a44bv405bllgjgd632ngjgc0i3q9n9im-review-shell.drv' failed to build
```

<details/>

ambroisie commented:
@litchipi we can bump the version of the package, as [it is now up to v1.1.0](https://github.com/mealie-recipes/mealie/releases/tag/v1.1.0).

litchipi commented:
@ambroisie Yep I'm on it, I hope I'll be able to reproduce any error locally !

ambroisie commented:
@litchipi for the `pydantic` issue, I think the cleanest option would be to open a PR upstream to update their code to v2 and accommodate its breaking changes. We could then `fetchpatch` the PR for packaging purposes.

Let me know if that gives you trouble, and I'll work on it when I have cycles to spare.

litchipi commented:
@ambroisie Because of the migration to `pydantic` v2, it breaks everything, a lot of the tests, and it seams that `mealie` just doesn't work anymore. At least I couldn't `poetry run pytest` on it directly.
I'm doing the changes now, and will propose a PR upstream, but I may want to put the changes as a patch file for now as the rest may take a while to get sorted, reviewed, cleaned, merged.

Also as I am not a contributor to `mealie`, there are some things going wrong in the code that is really difficult for me to debug, so I may need some external help on this.
For now I already have some things, you can check https://github.com/litchipi/mealie on branches `fixup_mealie_next` and `fixup_v1_1_0`

Atemu commented on nixos/modules/services/web-apps/mealie.nix:
```diff
@@ -0,0 +1,79 @@
+      description = "Port on which to serve the Mealie service.";
+    };
+
+    extraConfig = lib.mkOption {
```
As per RFC42, please turn this into a freeform `settings` option.

Atemu commented on pkgs/by-name/me/mealie/package.nix:
```diff
@@ -0,0 +1,103 @@
+{ lib, callPackage, fetchFromGitHub, nixosTests, python3, python3Packages, writeShellScript }: let
+  version = "1.0.0-RC2";
```
Nit:

```suggestion
{ lib, callPackage, fetchFromGitHub, nixosTests, python3, python3Packages, writeShellScript }:

let
  version = "1.0.0-RC2";
```

There are no clear rules on this yet but this is pretty ugly and not done elsewhere in Nixpkgs.

Atemu commented on pkgs/by-name/me/mealie/package.nix:
```diff
@@ -0,0 +1,103 @@
+  format = "pyproject";
+
+  patches = [
+    ./mealie_statedir.patch
+    ./mealie_init_db.patch
+    ./mealie_logger.patch
```
Please add a comment with info on what these do, why they're necessary and why they're not upstream.

mweinelt replied:
In fact, please bring these upstream.

ambroisie commented:
@Atemu I dont' have merge bits (I assume that's why) I can't mark them as resolved...

litchipi commented:
@Atemu thanks for the review ! I addressed your remarks
@ambroisie I got confused in what was wrong, I just had to override `pydantic` with the legacy `pydantic_1` version as the upstream mealie didn't make the change yet.

ambroisie commented on pkgs/by-name/me/mealie/package.nix:
```diff
@@ -0,0 +1,111 @@
+
+  pythonRelaxDeps = true;
+
+  propagatedBuildInputs = with python3Packages; [
+    aiofiles
+    alembic
+    aniso8601
+    appdirs
+    apprise
+    bcrypt
+    extruct
+    (fastapi.override {
+      pydantic = pydantic_1;
+    })
+    gunicorn
+    html2text
+    jinja2
+    lxml
+    orjson
+    passlib
+    pillow
+    psycopg2
+    pyhumps
+    pytesseract
+    python-dotenv
+    python-jose
+    python-ldap
+    python-multipart
+    python-slugify
+    pyyaml
+    rapidfuzz
+    recipe-scrapers
+    sqlalchemy
+    tzdata
+    uvicorn
+  ];
```
From what I've seen, the usual way to go about overriding some of the python packages for a `buildPythonPackage` is to use `packageOverrides`, see [`tandoor-recipes`](https://github.com/NixOS/nixpkgs/blob/f4a8d6d5324c327dcc2d863eb7f3cc06ad630df4/pkgs/applications/misc/tandoor-recipes/default.nix#L8) for example.

So something like the following should work:

```nix
  python = python3.override {
    packageOverrides = self: super: {
      pydantic = self.pydantic_1;
    };
  };
```

And do `passthru = { inherit python; }` as well.

ambroisie replied:
Ah and of course, make sure you use `propagatedBuildInputs = with python.pkgs; [ ... ]` to use those overrides.

ambroisie commented on pkgs/by-name/me/mealie/package.nix:
```diff
@@ -0,0 +1,111 @@
+  ];
+
+  passthru = {
+    inherit python3;
```
I would rather it be named `python`, + see my other comment.

Atemu commented:
> I can't mark them as resolved...

Oh, I had assumed every maintainer can do that. I also had it the wrong way around; of course the author should mark as resolved as they address reviews.

> @Atemu I dont' have merge bits (I assume that's why) 

Let's fix that ;)

https://github.com/NixOS/nixpkgs/issues/50105#issuecomment-1919489831

Atemu commented on nixos/modules/services/web-apps/mealie.nix:
```diff
@@ -0,0 +1,82 @@
+    };
+
+    settings = lib.mkOption {
+      type = lib.types.attrs;
```
This isn't critical but note that RFC42 includes a more specific type for freeform options:

https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md#part-1-1

litchipi committed:
mealie: init at v1.2.0

Signed-off-by: Litchi Pi <litchi.pi@proton.me>

litchipi committed:
nixos/mealie: init module

Signed-off-by: Litchi Pi <litchi.pi@proton.me>

litchipi committed:
nixos/mealie: add to release notes 24.05

Signed-off-by: Litchi Pi <litchi.pi@proton.me>

litchipi committed:
nixos/mealie: init tests

Signed-off-by: Litchi Pi <litchi.pi@proton.me>

litchipi commented:
@Atemu The RFC was rather confusing I gotta admit, hope it's as you expected now. As the settings are passed using environment variables, I didn't use any `pkgs.format`.
@ambroisie Thanks for the catch, that'll be useful for me later on ;-)

litchipi commented:
Had to fix some other things as well, it should now be ready to merge :slightly_smiling_face:

h7x4 commented on pkgs/by-name/me/mealie/package.nix:
```diff
@@ -0,0 +1,116 @@
+      database by providing the URL and Mealie will automatically import the relevant data or add a family recipe with
+      the UI editor.
+    '';
+    homepage = "https://nightly.mealie.io";
```
```suggestion
    homepage = "https://nightly.mealie.io";
    changelog = "https://github.com/mealie-recipes/mealie/releases/tag/${src.rev}"
```

Atemu replied:
Why not simply

```suggestion
    homepage = "https://mealie.io";
```

h7x4 commented on pkgs/by-name/me/mealie/package.nix:
```diff
@@ -0,0 +1,116 @@
+    '';
+    homepage = "https://nightly.mealie.io";
+    license = licenses.agpl3Only;
+    maintainers = with maintainers; [ litchipi ];
```
```suggestion
    maintainers = with maintainers; [ litchipi ];
    mainProgram = "start-mealie";
```

h7x4 reviewed (commented):
Here's a few suggestions :)

ambroisie commented on pkgs/by-name/me/mealie/mealie-frontend.nix:
```diff
@@ -0,0 +1,46 @@
+  src = "${src}/frontend";
+
+  yarnOfflineCache = fetchYarnDeps {
+    yarnLock = "${src}/frontend/yarn.lock";
```
No this is fine, as long as your specify the hash.

litchipi commented:
@h7x4 Thanks for the review, made the necessary changes ! :slightly_smiling_face:

Atemu commented on pkgs/by-name/me/mealie/package.nix:
```diff
@@ -0,0 +1,123 @@
+      database by providing the URL and Mealie will automatically import the relevant data or add a family recipe with
+      the UI editor.
+    '';
+    homepage = "https://nightly.mealie.io";
```
Why not simply

```suggestion
    homepage = "https://mealie.io";
```

Atemu commented on nixos/modules/services/web-apps/mealie.nix:
```diff
@@ -0,0 +1,89 @@
+        User = "mealie";
+        ExecStartPre = "${pkg.python.interpreter} ${pkg}/bin/init_db.py";
+        ExecStart = "${pkg}/bin/start-mealie -b ${cfg.listenAddress}:${builtins.toString cfg.port}";
+        EnvironmentFile = lib.optional (cfg.credentialsFile != null) cfg.credentialsFile;
```
It's probably better to not set the option at all than to an empty list.

```suggestion
        EnvironmentFile = lib.mkIf (cfg.credentialsFile != null) cfg.credentialsFile;
```

Atemu commented on pkgs/by-name/me/mealie/package.nix:
```diff
@@ -0,0 +1,123 @@
+      --replace 'script_location = alembic' 'script_location = ${src}/alembic'
+
+    rm -f $out/bin/*
+    ln -s $out/${python.sitePackages}/mealie/db/init_db.py $out/bin/init_db.py
```
The init_db script should probably not be in bin.

I'd put it in libexec, wrap it with the interpreter and expose it via a passtrhu attr.

litchipi replied:
Fair enough, I just didn't find a way to wrap it properly (as it's dependent on some Python imports relative to script path).
Sorry I'll need an example of how to do this

mweinelt replied:
Use `makeWrapper` and set the `PYTHONPATH`.

mweinelt commented on nixos/modules/services/web-apps/mealie.nix:
```diff
@@ -0,0 +1,89 @@
+  config = lib.mkIf cfg.enable {
+    systemd.services.mealie = {
+      description = "Mealie, a self hosted recipe manager and meal planner";
+      after = [
+        "network.target"
+        "network-online.target"
+      ];
```
```suggestion
      after = [
        "network-online.target"
      ];
```

The latter supersedes the former.

mweinelt commented on nixos/modules/services/web-apps/mealie.nix:
```diff
@@ -0,0 +1,89 @@
+    };
+
+    settings = lib.mkOption {
+      type = lib.types.submodule {
+        freeformType = with lib.types; attrsOf str;
+      };
```
This looks like a flat attribute set, so `attrsOf str` should be sufficient.

litchipi replied:
I tried to apply what was requested here https://github.com/NixOS/nixpkgs/pull/278454#discussion_r1473151732
Isn't it part of the RFC to do so ?

mweinelt replied:
The benefit of using the freeformType is, that you can add defaults (e.g. the `port` option in that example) on top of `settings`. if you don't plan on doing that, there is little point in going for this complexity.

ambroisie replied:
`attrsOf str` is overly restrictive IMO, given that we `mapAttrs (_: toString)` later.

litchipi replied:
Changed back to `attrsOf anything` so we can use numbers and booleans

mweinelt commented on nixos/modules/services/web-apps/mealie.nix:
```diff
@@ -0,0 +1,89 @@
+      environment = {
+        PYTHONPATH = pkg.python.pkgs.makePythonPath ([ pkg ] ++ pkg.propagatedBuildInputs);
+        PRODUCTION = "true";
+        MEALIE_LOG_FILE = "/var/log/mealie/mealie.log";
```
How much logging happens here? Without log rotation this can be annoying. Ideally logging would go into the journal.

litchipi replied:
Actually, the service logs everything into a file (hardcoded path in the code), so not in the journal.
I only set this env var in order to redirect logs into it.

mweinelt replied:
Ideally there would be some sort of log rotation going on, so it doesn't accumulate logs forever.

Atemu replied:
Cases like this can usually be worked around by

```suggestion
        MEALIE_LOG_FILE = "/dev/stdout";
```

litchipi replied:
That unfortunately doesn't work,  `nixosTests.mealie` returns "/dev/stdout not found".
Is there some kind of protection for `/dev/` in systemd services ?
Because with `StandardOutput="journal"` that would be the ideal setup for log rotation

litchipi replied:
However, as I removed `LOGGER_FILE` entirely, I also had to remove some admin features that was fetching logs from the file.

For now, let's keep the "dirty" removal of all of it, as stated in https://github.com/mealie-recipes/mealie/pull/3102#discussion_r1476531794, the logging will be refactored soon, so I'll adapt it to the final form then

mweinelt commented on pkgs/by-name/me/mealie/package.nix:
```diff
@@ -0,0 +1,123 @@
+    uvicorn
+  ];
+
+  doCheck = true;
+
```
```suggestion
```

Always defaults to true in python builds.

mweinelt commented on pkgs/by-name/me/mealie/package.nix:
```diff
@@ -0,0 +1,123 @@
+  doCheck = true;
+
+  postInstall = let
+    start_script = writeShellScript "start-mealie" ''
+      export STATIC_FILES="${frontend}"
+      ${lib.getExe python.pkgs.gunicorn} "$@" -k uvicorn.workers.UvicornWorker mealie.app:app;
+    '';
```
I think `makeWrapper` would make a better fit here. Also, why call it `start-mealie` and not just `mealie`? Executing a file is starting it, so that much is implied.

litchipi replied:
I called it `start-mealie` because it's not the mealie "binary", but just a script whose job is to start mealie.
That being said, I can rename it if you want, not a problem for me

mweinelt replied:
If this is they way through which mealie gets started, I think it is fair to call it mealie. Using `makeWrapper`, you could also set the `PYTHONPATH` correctly, so it wouldn't need to be passed in the systemd unit.

litchipi commented:
@mweinelt I cannot push the patches upstream, they use hardcoded path (because they expect to be inside a docker container), this is what the patches are about, make the hardcoded path use env vars instead in order to make it configurable.

mweinelt commented:
Yeah, the question is, whether the upstream would consider allowing overrides, like `os.getenv("SOME_VAR", "actual_default")`, that we could rely on. I think this is a matter of talking to the upstream, whether they can accomodate our use case.

litchipi commented:
@mweinelt Alright I'll propose the changes upstream. In the meantime, let's keep the patches here, and I'll clean them up in a following PR !

Atemu commented:
@litchipi that'd be great. Make sure you add a link above the patches so that others can check their status.

Atemu commented on nixos/modules/services/web-apps/mealie.nix:
```diff
@@ -0,0 +1,86 @@
+  config = lib.mkIf cfg.enable {
+    services.logrotate.settings.mealie = {
+      enable = true;
+      files = [ "/var/log/mealie/mealie.log" ];
```
If they user is able to configure the logfile position via `settings`, this should probably use the same value.

```suggestion
      files = [ (settings.MEALIE_LOG_FILE or "/var/log/mealie/mealie.log") ];
```

Have you tried whether the `/dev/stdout` hack works? That'd be preferable to log rotation I think.

litchipi replied:
I saw this, but never tried it (was really poking randomly), I'll attempt something, tell me if that could work

litchipi replied:
Got a solution working: See https://github.com/NixOS/nixpkgs/pull/278454#discussion_r1477017845

Atemu commented on pkgs/by-name/me/mealie/package.nix:
```diff
@@ -0,0 +1,129 @@
+    uvicorn
+  ];
+
+  start_script = writeShellScript "start-mealie" ''
+    ${lib.getExe python.pkgs.gunicorn} "$@" -k uvicorn.workers.UvicornWorker mealie.app:app;
+  '';
```
Please use a let bindings rather than adding attrs to the drv.

Atemu commented on pkgs/by-name/me/mealie/package.nix:
```diff
@@ -0,0 +1,129 @@
+  ];
+
+  passthru = {
+    init_db = pkg: writeShellScript "init-mealie-db" ''
+      export PYTHONPATH=${python.pkgs.makePythonPath ([pkg] ++ propagatedBuildInputs)}
+      ${python.interpreter} ${pkg}/${python.sitePackages}/mealie/db/init_db.py
+    '';
```
Why is this a function? Make this a regular attr.

You can use `finalAttrs.finalPackage` to refer to the package itself from itself.

litchipi replied:
Ok I missed that last piece of info thanks :sweat_smile:

Atemu replied:
Oh indeed, it cannot do that. Sorry for the suggestion, I had assumed that it could. In that case you'd have to pass the package to itself or simply make this an executable in `libexec` as I initially recommended.

litchipi replied:
I made a `init_db` script stored inside `$out/libexec/init_db`, I hope that's what you meant :smile:

Atemu reviewed (approved):
Appears to work as intended, including NLP. Great work! :)

One super minor thing I noticed is that the version the app reports is `develop`. That's probably some build flag where you need to pass the version.

litchipi commented:
Actually that's hardcoded:
- https://github.com/mealie-recipes/mealie/blob/42a33cd993b4638f181d4083a7c5d21bf110e4e4/mealie/__init__.py#L1
- https://github.com/mealie-recipes/mealie/blob/mealie-next/mealie/core/settings/static.py#L5C1-L5C12

And I see no way to override this without a patch.. That may be worth a PR on the repository, that would get it from env var

ambroisie commented:
> And I see no way to override this without a patch.

IMO it's fine to `substitueInpPlace` during `postPatch` to patch it.

litchipi commented:
Changed it on the package !
Version `1.2.0` is out now, will do the version upgrade soon, hope it won't be too much of a hassle

litchipi commented:
Change was trivial, updated to latest `1.2.0` version, and rebased over latest `master` :slightly_smiling_face:

ambroisie commented on pkgs/by-name/me/mealie/mealie-frontend.nix:
```diff
@@ -0,0 +1,46 @@
+src: version:
+{ lib, fetchYarnDeps, nodejs_18, prefetch-yarn-deps, stdenv, ... }: stdenv.mkDerivation {
```
```suggestion
{ lib, fetchYarnDeps, nodejs_18, prefetch-yarn-deps, stdenv }: stdenv.mkDerivation {
```

ambroisie commented on pkgs/by-name/me/mealie/mealie-frontend.nix:
```diff
@@ -0,0 +1,46 @@
+    export HOME=$(mktemp -d)
+    yarn config --offline set yarn-offline-mirror "$yarnOfflineCache"
+    fixup-yarn-lock yarn.lock
+    command -v yarn
```
?

```suggestion
```

ambroisie commented on pkgs/by-name/me/mealie/package.nix:
```diff
@@ -0,0 +1,151 @@
+  pyproject = true;
+
+  patches = [
+    # See https://github.com/mealie-recipes/mealie/pull/3102
+    # Replace hardcoded paths in code with environment variables (meant for inside Docker only)
+    # So we can configure easily where the data is stored on the server
+    ./mealie_model_path.patch
+    ./mealie_init_db.patch
```
Now that you've upstreamed the patches: it's idiomatic to use `fetchpatch` to download the patches, rather than vendor them.

So something like:

```suggestion
    # https://github.com/mealie-recipes/mealie/pull/3102
    # Replace hardcoded paths in code with environment variables (meant for inside Docker only)
    # So we can configure easily where the data is stored on the server
    (fetchpatch {
      name = "mealie-model-path.patch";
      url = "https://github.com/mealie-recipes/mealie/commit/e445705c5d26b895d806b96b2f330d4e9aac3723.patch";
      hash = ""; # FIXME
    })
    (fetchpatch {
      name = "mealie-init-db.patch";
      url = "https://github.com/mealie-recipes/mealie/commit/06c528bfac0708af66aa0629f2e2232ddf07768f.patch";
      hash = ""; # FIXME
    })
```

And delete those two files.

ambroisie reviewed (changes requested):
Small nits.

SuperSandro2000 commented on pkgs/by-name/me/mealie/package.nix:
```diff
@@ -0,0 +1,165 @@
+
+  crfpp = stdenv.mkDerivation {
+    pname = "mealie-crfpp";
+    version = "master";
```
This should be in the unstable-XXXX-XX-XX where the X's are the date

litchipi replied:
I'm sorry I don't understand, I should add `crfpp` to the whole nixpkgs packages ?
It's something fairly unmaintained, and here forked specifically for `mealie`, I don't really know if that's a wise choice

Atemu replied:
If this is a package others might rely on, yes. If this is a "vendored" dependency, it's fine to keep it in here. Though as @SuperSandro2000 said, please use the standard version format used for things built from git revisions rather than tags.

litchipi replied:
Ok I see, fixed

SuperSandro2000 commented on pkgs/by-name/me/mealie/mealie-logs-to-stdout.patch:
```diff
@@ -0,0 +1,103 @@
+ 
+ from .config import get_app_settings  # noqa E402
+ 
+-LOGGER_FILE = DATA_DIR.joinpath("mealie.log")
```
Would it be enough to replace the file here with /dev/stdout or /dev/log?

Atemu replied:
This was confirmed to not work previously,

SuperSandro2000 commented on pkgs/by-name/me/mealie/mealie-frontend.nix:
```diff
@@ -0,0 +1,45 @@
+    runHook preInstall
+    mv dist $out
+    runHook postInstall
+  '';
```
meta is missing

litchipi replied:
Added some metadata

Atemu commented on pkgs/by-name/me/mealie/package.nix:
```diff
@@ -28,7 +28,7 @@ let
   crfpp = stdenv.mkDerivation {
     pname = "mealie-crfpp";
-    version = "master";
+    version = "unstable-2024-02-12";
```
The date should be the commit date.

```suggestion
    version = "unstable-2021-10-10";
```

litchipi replied:
Ah yes sorry, good catch

Atemu replied:
You accidentally reverted this.

Atemu commented:
> If this PR is OK for you, let's merge this with the patches for now, and when a new release is made, at the same time as bumping the version I'll remove the patches ?

Yup, that's the way to go.

litchipi commented:
@mweinelt Does this PR looks good for you ? :smiley:

mweinelt commented on pkgs/by-name/me/mealie/mealie-frontend.nix:
```diff
@@ -0,0 +1,51 @@
+src: version:
+{ lib, fetchYarnDeps, nodejs_18, prefetch-yarn-deps, stdenv }: stdenv.mkDerivation {
```
Is pinning the nodejs version necessary?

mweinelt commented on pkgs/by-name/me/mealie/package.nix:
```diff
@@ -0,0 +1,160 @@
+, fetchFromGitHub
+, fetchpatch
+, makeWrapper
+, python3
```
Use `python3Packages` instead. Apparently it works better for cross compilation. It maps to `python3.pkgs`.

litchipi replied:
I do not only need `python3.pkgs`, I also require `python3.sitePackages`, to do an override of `packageOverrides` for pydantic, and call `python3.interpreter`

mweinelt replied:
Then use `python3Packages.python.sitePackages`.

litchipi replied:
Thanks, fixed it

mweinelt commented on pkgs/by-name/me/mealie/package.nix:
```diff
@@ -0,0 +1,160 @@
+      owner = "mealie-recipes";
+      repo = "crfpp";
+      rev = "c56dd9f29469c8a9f34456b8c0d6ae0476110516";
+      sha256 = "sha256-XNps3ZApU8m07bfPEnvip1w+3hLajdn9+L5+IpEaP0c=";
```
Generally, prefer hash over sha256, especially with SRI hashes.

mweinelt commented on nixos/doc/manual/release-notes/rl-2405.section.md:
```diff
@@ -81,6 +81,8 @@ The pre-existing [services.ankisyncd](#opt-services.ankisyncd.enable) has been m
 
 - [systemd-lock-handler](https://git.sr.ht/~whynothugo/systemd-lock-handler/), a bridge between logind D-Bus events and systemd targets. Available as [services.systemd-lock-handler.enable](#opt-services.systemd-lock-handler.enable).
 
+- [Mealie](https://nightly.mealie.io/), a self-hosted recipe manager and meal planner with a RestAPI backend and a reactive frontend application built in NuxtJS for a pleasant user experience for the whole family. Available as [services.mealie](#opt-services.mealie.enable)
```
[REST](https://en.wikipedia.org/wiki/REST) is a fixed term. So you can say "REST API" instead of "RestAPI backend".

litchipi replied:
Copy-pasted from the website, so I should maybe tell this to the creator as well :sweat_smile:

mweinelt reviewed (commented):
Only small stuff, thanks for asking.

litchipi commented:
@mweinelt Regarding https://github.com/NixOS/nixpkgs/pull/278454#discussion_r1499187540

The `StateDirectory` is used for `DATA_DIR` env var, the logs isn't used anymore, so I removed it

litchipi commented:
Made the changes and created clean commits

Atemu commented:
I think we're now at the point where all major issues have been ironed out. Any further improvements can come in follow-up PRs.

Thank you so much for sticking with us and pulling through @litchipi. Not everyone contributes such a high quality PR as their first contribution.
